Gazebo7 fixes

scpeters opened the pull request:
This fixes three separate build errors when using gazebo7 (related to #375):
- a27311b: a boost header was removed from a gazebo header and was being used implicitly by gazebo_ros_joint_state_publisher.cpp
- ce0ab10: resolve ambiguity between overloaded functions ( [upstream issue](https://bitbucket.org/osrf/gazebo/issues/1830) ).
- 194ebf8: sensor pointers changed from boost to std shared pointers. I've added a macro called `GAZEBO_SENSORS_USING_DYNAMIC_POINTER_CAST` to `gazebo_ros_utils.h` that provides a `using` statement so that `dynamic_pointer_cast` can be used more easily.

scpeters committed:
Fix gazebo7 build errors

The SensorPtr types have changed from boost:: pointers
to std:: pointers,
which requires boost::dynamic_pointer_cast to change to
std::dynamic_pointer_cast.
A helper macro is added that adds a `using` statement
corresponding to the correct type of dynamic_pointer_cast.
This macro should be narrowly scoped to protect
other code.

scpeters committed:
Add missing boost header

Some boost headers were remove from gazebo7 header files
and gazebo_ros_joint_state_publisher.cpp was using it
implicitly.

scpeters committed:
Fix compiler error with SetHFOV

In gazebo7, the rendering::Camera::SetHFOV function
is overloaded with a potential for ambiguity,
as reported in the following issue:
https://bitbucket.org/osrf/gazebo/issues/1830
This fixes the build by explicitly defining the
Angle type.

hsu commented on gazebo_plugins/include/gazebo_plugins/gazebo_ros_utils.h:
```diff
@@ -41,6 +41,14 @@
 
+#ifndef GAZEBO_SENSORS_USING_DYNAMIC_POINTER_CAST
+# if GAZEBO_MAJOR_VERSION >= 7
+#define GAZEBO_SENSORS_USING_DYNAMIC_POINTER_CAST using std::dynamic_pointer_cast
```
just for my own education, what's the reasoning behind using `using` as opposed to making `xxx::dynamic_pointer_cast` a macro?

scpeters replied:
In urdfdom we made [wrapper functions in types.h](https://github.com/ros/urdfdom_headers/blob/master/urdf_model/include/urdf_model/types.h#L72-L95) so you could call `urdf::dynamic_pointer_cast`. In this case, we haven't switched all of gazebo's pointers yet (sensors uses `std::` in gazebo7 while physics still uses `boost::`). It seemed like these `using` macros (which can be scoped) would be a more compact way to allow for different pointer types.

hsu commented:
other than the comment about the macro, everything looks good and compiles. Thanks!

hsu commented:
+1

scpeters committed:
gazebo_ros_utils.h: include gazebo_config.h

Make sure to include gazebo_config.h,
which defines the GAZEBO_MAJOR_VERSION macro
